core/src/main/java/ai/timefold/solver/core/impl/score/stream/bavet/BavetConstraintFactory.java

```diff
@@ -95,6 +101,28 @@ public <Stream_ extends BavetAbstractConstraintStream<Solution_>> Stream_ share(
 
     private <A> @NonNull UniConstraintStream<A> forEachForCriteria(@NonNull Class<A> sourceClass,
             ForEachFilteringCriteria criteria) {
+        // Required for node sharing, since using a lambda will create different instances
```

Comment:
The method is already complex enough even without long inner type declarations. Let's move it outside of the method. (Would keep it in the class though.)